### ISSUE TITLE
Linux libportaudio.so pre-built library does not appear to support ALSA

### ISSUE THREAD

**Issue by amberwhitehead** [csukuangfj/PortAudioSharp2#14]
This is a cool project, love it!

On Linux (Ubuntu 23.10) I have trouble getting any sound. Package installs successfully, API functions seem to work. But I get a `PortAudio.DeviceCount` of `0`. There is no `PortAudio.DefaultOutputDevice`, it's just `-1`.

After debugging a bit, I think that the issue is that the pre-built `libportaudio.so` for Linux is built without ALSA or Jack support, or the support isn't working. This is a problem because modern Linux systems expect to use ALSA or Jack for audio and not legacy OSS mode.

First try at a fix would be to recompile PortAudio for linux and update the file `scripts/libportaudio.so`. The compile needs to be done in an environment with ALSA dev package available. On Ubuntu that is `sudo apt install libasound2-dev`. The PortAudio build should detect ALSA and include it in the library. Similarly for Jack, the package would be `sudo apt install libjack-dev`.

There are probably considerations about the LIBC versioning when distributing binary libraries, as well as ALSA and Jack dependencies. I believe ALSA can be statically linked to avoid linking dependencies. That might be the easiest, just statically link ALSA into `scripts/libportaudio.so`.

**Comment by amberwhitehead** [comment]
In case it is helpful, I'm playing around with building PortAudio using `podman` with the following `Dockerfile`:

```
# Dockerfile
#
# Build PortAudio v19.7.0, Linux
#
# Use this file in an empty directory with commands:
#
#     podman build . -t portaudio-build
#     mkdir out
#     podman run --mount type=bind,source=out,dst=/out -t portaudio-build
#
# That should produce shared library in out/ directory.
# Using Ubuntu 16.04 means it hopefully will work in most modern Linux systems.

FROM ubuntu:16.04

RUN apt update
RUN apt install -y build-essential cmake git libasound2-dev libjack-dev
RUN git clone https://github.com/PortAudio/portaudio.git --branch v19.7.0 --single-branch
RUN mkdir build && cd build && cmake -DCMAKE_BUILD_TYPE=Release ../portaudio
RUN echo "#!/bin/bash\ncd /build && make -j12 && cp libportaudio.so /out" >> /script.sh
RUN chmod a+x /script.sh
ENTRYPOINT /script.sh
```

**Comment by csukuangfj** [comment]
Thanks!

Does it work for you after you replace the provided libportaudio.so with yours that has alsa support?

**Comment by amberwhitehead** [comment]
Yes, I got it working with the updated `libportaudio.so`. I tested by doing `dotnet add package PortAudioSharp2`, then doing `dotnet build` on my test program. Then I did the copy:

```
cp ../../portaudio/docker/out/libportaudio.so bin/Debug/net8.0/runtimes/linux-x64/native/libportaudio.so
```
Then I had to run the test program manually:
```
./bin/Debug/net8.0/Demo
```
I normally use `dotnet run`, but that way notices that the shared library has changed and copies it back.

With the updated shared library I get 20 audio devices found instead of 0.
